Add option to disable streaming to stdout file

AlanCoding opened the pull request:
This addresses https://github.com/ansible/ansible-runner/issues/926

My current intention is to make use of the setting (set it to True) in AWX, unconditionally. We don't want the additional disk writes this incurs. If the executor team is okay with the looks of this, I may ask @amolgautam25 to make that PR.

This use case is only concerned with the `pexpect` runner mode. I only messed with `subprocess` here for completeness. The subprocess mode also appears to somehow be connected with the introduction of this behavior. Maybe @ganeshrn has more history to share about this.

AlanCoding committed:
Add option to disable streaming to stdout file

AlanCoding committed:
Add a quick line to document new setting

john-westcott-iv commented on docs/intro.rst:
```diff
@@ -139,7 +139,8 @@ The **settings** file is a little different than the other files provided in thi
 
-* ``suppress_ansible_output``: ``False`` Allow output from ansible to not be printed to the screen
+* ``suppress_output_file``: ``False`` Allow output from ansible to not be streamed to the ``stdout`` or ``stderr`` files inside of the artifacts directory.
+* ``suppress_ansible_output``: ``False`` Allow output from ansible to not be printed to the screen.
```
Should there be a check to make these mutually exclusive? If they are both set ansible would "just run" with no output, maybe we don't care?

AlanCoding replied:
AWX will set them both to `False` by design. The output is processed by the callback receiver. If someone really wanted the standard out file (which is normally deleted at the end of the run anyway) then we can introduce a setting for them to get it back.

There's a feature in AWX to download the standard out file from a job run. It is illustrative that this works by a completely unrelated method. In the request-response cycle, it is written to disk from the database data and then delivered to the client. Then deleted.

AlanCoding committed:
Test refinement for feature to skip writing stdout

Do a quick poll until assertion is satisfied
Use iterate_timeout test util from review comment
reuse poll for the stdout file because of race condition
Skip test for pexpect because it does not work

AlanCoding commented on test/unit/test_runner.py:
```diff
@@ -31,16 +33,6 @@ def rc(request, tmp_path):
-    # about being nice
-    m = mocker.patch('time.sleep')
-    m.start()
-    request.addfinalizer(m.stop)
```
I checked the timing, and the whole module takes about 3.6 seconds. Maybe this was reducing it to 3.5 or 3.4, but the difference also could have been indistinguishable. I'm using `time.sleep` to make the assertion reliable, so that's why I did this.

Shrews commented on test/unit/test_runner.py:
```diff
@@ -152,3 +144,32 @@ def test_status_callback_interface(rc, mocker):
+    stdout_path = Path(rc.artifact_dir) / 'stdout'
+    # this can be subject to a race condition so we will be patient with the check
+    start_time = time.time()
+    while time.time() - start_time < 2.0:
```
Use `iterate_timeout()` here from `test/utils/common.py`. You can find examples in the test code.

Shrews reviewed (changes requested):
Looks pretty good to me, other than the comment on the tests to re-use the `iterate_timeout()` function.

Shrews commented on test/unit/test_runner.py:
```diff
@@ -154,13 +155,9 @@ def test_stdout_file_write(rc, runner_mode):
     # this can be subject to a race condition so we will be patient with the check
-    start_time = time.time()
-    while time.time() - start_time < 2.0:
+    for _ in iterate_timeout(2.0, 'stdout file to be written', interval=0.2):
```
We can get some pretty slow test VMs, so 2 seconds can be cutting it pretty darn close, which is likely why you see the test failure. I see no reason not to just bump this to, say, 30 seconds.

AlanCoding replied:
Thanks. The first time I checked results I saw 2 configurations failing and after adding the 2 second poll 1 was failing. So this would fit the picture that it's slow, dependent on the hardware.

AlanCoding commented:
Interesting... I'm wondering if this was a pre-existing failure of the same type

```
=================================== FAILURES ===================================
[31m[1m_____________________________ test_password_prompt _____________________________[0m
[gw3] linux -- Python 3.8.12 /home/zuul/src/github.com/ansible/ansible-runner/.tox/py38/bin/python

rc = <ansible_runner.config.runner.RunnerConfig object at 0x7ff864f2e520>

    def test_password_prompt(rc):
        rc.command = [sys.executable, '-c' 'import time; print(input("Password: "))']
        rc.expect_passwords[re.compile(r'Password:\s*?$', re.M)] = '1234'
        status, exitcode = Runner(config=rc).run()
        assert status == 'successful'
        assert exitcode == 0
        with open(os.path.join(rc.artifact_dir, 'stdout')) as f:
>           assert '1234' in f.read()
[1m[31mE           AssertionError: assert '1234' in ''[0m
[1m[31mE            +  where '' = <built-in method read of _io.TextIOWrapper object at 0x7ff864d44ee0>()[0m
[1m[31mE            +    where <built-in method read of _io.TextIOWrapper object at 0x7ff864d44ee0> = <_io.TextIOWrapper name='/tmp/pytest-of-zuul/pytest-0/popen-gw3/test_password_prompt0/artifacts/21213e93-55b2-4899-8dc7-edeebb7c59ec/stdout' mode='r' encoding='UTF-8'>.read[0m

exitcode   = 0
f          = <_io.TextIOWrapper name='/tmp/pytest-of-zuul/pytest-0/popen-gw3/test_password_prompt0/artifacts/21213e93-55b2-4899-8dc7-edeebb7c59ec/stdout' mode='r' encoding='UTF-8'>
rc         = <ansible_runner.config.runner.RunnerConfig object at 0x7ff864f2e520>
status     = 'successful'

[1m[31mtest/integration/test_runner.py[0m:22: AssertionError
```

I can apply the same methods to that assertion. That looks like it may be the proper thing to do.

Shrews commented:
The disk space issue aside (no idea why we're seeing that on occasion), each time the new `test_stdout_file_write` has failed has been with the `pexpect` path (not the `subprocess` path). That may be worth looking into.

Shrews commented:
> The disk space issue aside (no idea why we're seeing that on occasion), each time the new `test_stdout_file_write` has failed has been with the `pexpect` path (not the `subprocess` path). That may be worth looking into.

FYI, the failure using `pexpect` is reproducible on Linux. It can take several runs, but definitely reproducible.

Shrews reviewed (changes requested):
Removing approval until the test failure under `pexpect` is figured out.

AlanCoding commented:
Would it be a problem for me to do a `pytest.skip`  for the pexpect case? It looks like it's just buggy from some problem with the library maybe.

AlanCoding commented:
@Shrews I added the skip, which looks like it's working to stabilize tests. I can make an issue to cover the stdout log file not working with pexpect

Shrews reviewed (approved):
It makes me a sad panda that we have to skip the test case we _explicitly_ want this feature for, but this change obviously didn't create the error, only exposed it. So I guess skipping it is ok for now.
